python/tskit/stats.py

```diff
@@ -163,3 +167,457 @@ def r2_matrix(self):
+        last_block = np.zeros(tree.tree_sequence.num_nodes, "int64") - 1
+        num_record = 0
+        while tree != tskit.NULL:
+            if tree.interval.right > left:
+                current_block = tree_blocks[tree.index - tree_offset]
+                nodes_in_tree = np.array(
+                    [i for i in tree.nodes() if tree.is_internal(i)]
+                )
+                nodes_to_add = nodes_in_tree[
+                    np.where(last_block[nodes_in_tree] != current_block)
+                ]
+                if len(nodes_to_add) > 0:
+                    idx = np.arange(num_record, num_record + len(nodes_to_add))
+                    last_block[nodes_to_add] = current_block
+                    indices[nodes_to_add] = idx
+                    if table_size < num_record + len(nodes_to_add):
+                        table_size += buffer_size
+                        time.resize(table_size)
+                        block.resize(table_size)
+                        weights.resize((table_size, self.num_weights))
+                    time[idx] = [tree.time(i) for i in nodes_to_add]
+                    block[idx] = current_block
+                    num_record += len(nodes_to_add)
+                weights[indices[nodes_in_tree], :] += running_weights[nodes_in_tree, :]
+
+            if tree.interval.right < right:
+                # if current tree does not cross window boundary, move to next
+                tree.next()
+                self._update_weights_by_edge_diff(
+                    tree, next(edge_diffs), running_weights
+                )
+            else:
+                # use current tree as initial tree for next window
+                break
+
+        return self.CoalescenceTimeTable(time, block, weights)
+
+    def _generate_ecdf_tables(self, ts, window_breaks):
+        """
+        Return generator for ECDF tables across genomic windows defined by
+        ``window_breaks``.
+
+        ..note:: This could be used in methods in place of loops over
+            pre-assembled tables.
+        """
+
+        tree = ts.first()
+        edge_diffs = ts.edge_diffs()
+        running_weights = np.zeros((ts.num_nodes, self.num_weights), "int64")
+        self._update_weights_by_edge_diff(tree, next(edge_diffs), running_weights)
+        for left, right in zip(window_breaks[:-1], window_breaks[1:]):
+            yield self._build_ecdf_table_for_window(
+                left, right, tree, edge_diffs, running_weights
+            )
+
+    def __init__(
+        self,
+        ts,
+        sample_sets=None,
+        weight_func=None,
+        window_breaks=None,
+        blocks_per_window=None,
+    ):
+
+        assert isinstance(ts, trees.TreeSequence)
+
+        if sample_sets is None:
+            self.sample_sets = [list(ts.samples())]
+        elif all([isinstance(i, (int, np.int32, np.int64)) for i in sample_sets]):
+            self.sample_sets = [sample_sets]
+        else:
```

Comment:
There's quite a bit of complexity here around the input types - maybe we could just make it something simple for now, and create an issue to track adding more complex behaviour later?